### DISCUSSION TITLE
Add Option For Copy Product Special Pricing when using Admin -> Copy Product

### DISCUSSION THREAD

**Discussion by neekfenwick** [zencart/zencart#6257] [Ideas]
Correct me if I'm wrong, but if you have a Special price set on a product, say discounting it from $20 to $18, then using the Copy Product button does not copy the special discount.

As I understand it, Specials are a core Zen Cart feature.  I'm creating about 15 new products based on existing ones across our shop and it's a bit of a pain to copy each one, and then have to go copy all the special pricing as well.

I understand this may not be the required behaviour, so I suggest adding a checkbox to the confirm box, similar to the "Copy Product Attributes" one that appears if the product has attributes.

Is there some other feature that would go well with this that I'm not aware of, for example any mass actions one can take on a product where specials would come into play?

I see that `zen_remove_product()` does delete from `TABLE_SPECIALS` so that's cool.

**Comment by torvista** [comment]
Your suggestion is logical to me.

One GUI enhancement I have thought of adding, is javascript-hiding all those duplicate options until duplicate is selected, for a cleaner look.

**Reply by neekfenwick** [reply]
Sorry @torvista, I'm probably being dense but what 'duplicate options' are you referring to?  The word 'duplicate' is mentioned a few times in the "Copy to" box that appears when you click 'C' to copy a product but that doesn't sound like what you're talking about.

**Reply by torvista** [reply]
all the stuff in the green box does not need to be shown until Duplicate is selected:
![2024-02-27 11_13_33-Admin Category Product Listing — Mozilla Firefox](https://github.com/zencart/zencart/assets/4391026/6a78e6e1-5419-4f9f-a4cf-2b2cfdae5ca6)

**Reply by neekfenwick** [reply]
Ah OK.  I've never used the 'Link' functionality, only ever dealing with very simple one-to-one category-product relationships. Thanks, I'll have a look.

**Reply by neekfenwick** [reply]
Where is your "URI Mapping" bit coming from?  It looks like it's being pulled in from an addon, via the notifier line at https://github.com/zencart/zencart/blob/119d28c91282c7bbb9e239bcbed152893d29e8e0/admin/includes/modules/copy_product.php#L56

My initial approach was to 'tag' each `$content` infoBox row that you have highlighted with a new CSS class like `hide-tag`, then have the Duplicate radio button handler show/hide each row with this tag class.  But, if infoBox rows are contributed by addons, they won't have this new tag class, so it's too brittle.  I could use a dumber approach, and just hide *all* rows on the page after the Duplicate radio button section and before the "Open duplicate product for editing?" checkbox.  However, this would break any addon that adds options to the page that apply to both linked and duplicate products i.e. those that should not be hidden when you select the "Link Product" radio, e.g. an addon that offers an option to modify the product being linked in some way at the same time as linking it.

It's a bit of a conundrum.  So, if we want this show/hide feature, it seems best to stick to my first plan of adding a tag class to each row that should be hidden, so that addon-contributed UI elements are left visible all the time.  Addons may then be updated to add this CSS tag class if they want to take advantage of it, if they know their feature is irrelevant to "Link product".

Does that make sense? Any feedback?

**Reply by torvista** [reply]
Yes, URI mapping and Kits are add-ons.
Your conclusion make the most sense, I always think add-ons need to adapt to core changes/should not constrict core advances.

**Comment by neekfenwick** [comment]
Here's a feature that's always bugged me, when Copying a product, the "link" option is checked by default, with "duplicate" available as an alternative. I personally have never once linked a product, so in our local store I have modified the file to pass `true` as the `$checked` parameter to `zen_draw_radio_field()` for the duplicate checkbox instead.
https://github.com/zencart/zencart/blob/82af1e1a04433941eea799416c2ebce271b23f51/admin/includes/modules/copy_product.php#L40-L45

Do you guys who manage more diverse stores generally Link, or Duplicate products?

I would very much like to add a config option to control the checked-by-default radio here, so you can configure your store so the Duplicate option is selected when the page opens.

**Reply by proseLA** [reply]
for my few clients, i have much more duplicated products than linked products...

**Comment by neekfenwick** [comment]
Please see PR #6262 
I have not implemented the config option mentioned above.  Can do, if approved.

**Comment by torvista** [comment]
You can't please everyone all the time (our products are massively linked).
If some default option irritates me enough I create a hidden constant in gID=6 and set it with whatever option was used last/no user intervention (like product/category sorting, despite it now being a session variable).

**Reply by neekfenwick** [reply]
Yeah, this sounds reasonable.  It would be a big job to have configurable defaults for absolutely everything.  I very rarely actually use the Copy Product button, we've got our own in-house generation system as we tend to make hundreds of products at a time.
